File: namui/src/render/Rect.ts
import { Color, InputRect } from "canvaskit-wasm";
import { AfterDraw } from "..";
import {
  MouseEventCallback,
  DrawCommand,
  RenderingTree,
  BorderPosition,
} from "../type";
import { nanoid } from "nanoid";

export function Rect({
  x,
  y,
  width,
  height,
  id,
  style: { stroke, fill, round },
  onClick,
  onClickOut,
  onMouseIn,
  onMouseMoveIn,
  onMouseMoveOut,
  onMouseDown,
  onMouseUp,
  onAfterDraw,
}: {
  x: number;
  y: number;
  width: number;
  height: number;
  id?: string;
  style: {
    stroke?: {
      color: Color;
      width: number;
      borderPosition: BorderPosition;
    };
    fill?: {
      color: Color;
    };
    round?: {
      radius: number;
    };
  };
  onClick?: MouseEventCallback;
  onClickOut?: MouseEventCallback;
  onMouseIn?: () => void;
  onMouseMoveIn?: MouseEventCallback;
  onMouseMoveOut?: MouseEventCallback;
  onMouseDown?: MouseEventCallback;
  onMouseUp?: MouseEventCallback;
  onAfterDraw?: (id: string) => void;
}): RenderingTree {
  const renderingTree = [];
  function getRectPath(rect: InputRect) {
    const rectPath = new CanvasKit.Path();
    if (round) {
      rectPath.addRRect(CanvasKit.RRectXY(rect, round.radius, round.radius));
    } else {
      rectPath.addRect(rect);
    }
    return rectPath;
  }

  let rect: InputRect;
  if (!stroke || stroke.borderPosition === BorderPosition.outside) {
    rect = CanvasKit.XYWHRect(x, y, width, height);
  } else if (stroke.borderPosition === BorderPosition.inside) {
    rect = CanvasKit.XYWHRect(
      x + stroke.width,
      y + stroke.width,
      width - 2 * stroke.width,
      height - 2 * stroke.width,
    );
  } else {
    rect = CanvasKit.XYWHRect(
      x + stroke.width / 2,
      y + stroke.width / 2,
      width - stroke.width,
      height - stroke.width,
    );
  }

  const rectPath = getRectPath(rect);

  const drawCommands: DrawCommand[] = [];

  if (stroke) {
    const strokePaint = new CanvasKit.Paint();
    strokePaint.setColor(stroke.color);
    strokePaint.setStrokeWidth(stroke.width);
    strokePaint.setStyle(CanvasKit.PaintStyle.Stroke);
    strokePaint.setAntiAlias(true);

    drawCommands.push({
      type: "path",
      path: rectPath,
      paint: strokePaint,
    });
  }

  if (fill) {
    const fillPaint = new CanvasKit.Paint();
    fillPaint.setColor(fill.color);
    fillPaint.setStyle(CanvasKit.PaintStyle.Fill);
    fillPaint.setAntiAlias(true);

    drawCommands.push({
      type: "path",
      path: rectPath,
      paint: fillPaint,
    });
  }

  if (onAfterDraw) {
    if (!id) {
      id = nanoid();
    }
    renderingTree.push(
      AfterDraw((param) => {
        onAfterDraw(id!);
      }),
    );
  }

  renderingTree.push({
    drawCalls: [
      {
        commands: drawCommands,
      },
    ],
    id,
    onClick,
    onClickOut,
    onMouseIn,
    onMouseMoveIn,
    onMouseMoveOut,
    onMouseDown,
    onMouseUp,
  });

  return renderingTree;
}

File: namui/src/engine/render/rect.rs
use crate::engine::{self, *};
pub enum BorderPosition {
    Inside,
    Outside,
    Middle,
}

pub struct RectStroke {
    pub color: Color,
    pub width: f32,
    pub border_position: BorderPosition,
}
pub struct RectFill {
    pub color: Color,
}
pub struct RectRound {
    pub radius: f32,
}
#[derive(Default)]
pub struct RectStyle {
    pub stroke: Option<RectStroke>,
    pub fill: Option<RectFill>,
    pub round: Option<RectRound>,
}
#[derive(Default)]
pub struct RectParam {
    pub x: f32,
    pub y: f32,
    pub width: f32,
    pub height: f32,
    pub id: Option<String>,
    pub style: RectStyle,
    pub on_click: Option<Box<dyn Fn(&engine::Xy<f32>)>>,
    //   pub onClickOut?: MouseEventCallback,
    //   pub onMouseIn?: () => void,
    //   pub onMouseMoveIn?: MouseEventCallback,
    //   pub onMouseMoveOut?: MouseEventCallback,
    //   pub onMouseDown?: MouseEventCallback,
    //   pub onMouseUp?: MouseEventCallback,
    //   pub onAfterDraw?: (id: string) => void,
}

pub fn rect(
    RectParam {
        x,
        y,
        width,
        height,
        style: RectStyle {
            stroke,
            fill,
            round,
        },
        id,
        on_click,
        ..
    }: RectParam,
) -> RenderingTree {
    let mut rendering_tree: Vec<RenderingTree> = vec![];
    let rect: engine::XywhRect<f32> = match stroke {
        None
        | Some(RectStroke {
            border_position: BorderPosition::Outside,
            ..
        }) => XywhRect {
            x,
            y,
            width,
            height,
        },
        Some(RectStroke {
            border_position: BorderPosition::Middle,
            width: stroke_width,
            ..
        }) => XywhRect {
            x: x + stroke_width,
            y: y + stroke_width,
            width: width - 2.0 * stroke_width,
            height: height - 2.0 * stroke_width,
        },
        Some(RectStroke {
            border_position: BorderPosition::Inside,
            width: stroke_width,
            ..
        }) => XywhRect {
            x: x + stroke_width / 2.0,
            y: y + stroke_width / 2.0,
            width: width - stroke_width,
            height: height - stroke_width,
        },
    };

    let rect_path = get_rect_path(rect, round);

    let mut draw_commands: Vec<DrawCommand> = vec![];

    if let Some(RectStroke {
        color,
        width: stroke_width,
        ..
    }) = stroke
    {
        let stroke_paint = engine::Paint::new();
        stroke_paint.set_color(&color);
        stroke_paint.set_stroke_width(stroke_width);
        stroke_paint.set_style(&engine::PaintStyle::Stroke);
        stroke_paint.set_anti_alias(true);

        draw_commands.push(DrawCommand::Path(PathDrawCommand {
            path: rect_path.clone(),
            paint: stroke_paint,
        }));
    };

    if let Some(RectFill { color }) = fill {
        let fill_paint = engine::Paint::new();
        fill_paint.set_color(&color);
        fill_paint.set_style(&engine::PaintStyle::Fill);
        fill_paint.set_anti_alias(true);

        draw_commands.push(DrawCommand::Path(PathDrawCommand {
            path: rect_path.clone(),
            paint: fill_paint,
        }));
    };

    // TODO
    //   if (onAfterDraw) {
    //     if (!id) {
    //       id = nanoid();
    //     }
    //     renderingTree.push(
    //       AfterDraw((param) => {
    //         onAfterDraw(id!);
    //       }),
    //     );
    //   }

    rendering_tree.push(RenderingTree::Node(RenderingData {
        draw_calls: vec![DrawCall {
            commands: draw_commands,
        }],
        id,
        on_click,
    }));

    RenderingTree::Children(rendering_tree)
}

//   function getRectPath(rect: InputRect) {
fn get_rect_path(rect: XywhRect<f32>, round: Option<RectRound>) -> engine::Path {
    let rect_path = engine::Path::new();

    if let Some(round) = round {
        rect_path.add_rrect(rect.into_ltrb(), round.radius, round.radius);
    } else {
        rect_path.add_rect(rect.into_ltrb());
    }

    rect_path
}
